splunk-adx-alert-addon/adx_connector.py

```diff
@@ -0,0 +1,155 @@
+                                                                                self.app_secret, self.tenant_id)
+
+        client = QueuedIngestClient(kcsb)
+        ingestion_props = IngestionProperties(
+            database=self.database_name,
+            table=self.table_name,
+            data_format=DataFormat.JSON,
+            report_level=ReportLevel.FailuresAndSuccesses,
+            ingestion_mapping_reference=self.table_mapping_name,
+            flush_immediately=False,
+            ingestion_mapping_kind=IngestionMappingKind.JSON
+        )
+        self.client = client
+        self.ingestion_props = ingestion_props
+        self.retry_times=5, 
+        self.retry_interval=5
+        self.exponential_backoff=True
+
+
+
+    def send(self, helper, events):
+        for event in events:
+            if helper.get_param("durable_mode") == "1" :
+                #durable mode ingestion : first we write to buffer and later we ingest them to ADX
+                try:
+                    timestr = time.strftime("%Y%m%d-%H%M")
+                    buffer_file_name = 'ADXIngest-' + timestr + '.json'
+                    parentdir = os.getcwd()
+                    buffer_folder_name = os.path.join(parentdir,"ADX_Ingest")
+                    if not os.path.isdir(buffer_folder_name):
+                        os.makedirs(buffer_folder_name)
+                    if not os.path.isfile(os.path.join(buffer_folder_name,buffer_file_name)):
+                        with open(os.path.join(buffer_folder_name,buffer_file_name), 'w+') as outfile:
+                            json.dump(event,outfile)
+                            self.successfully_sent_events_number += 1
+                    else:
+                        with open(os.path.join(buffer_folder_name,buffer_file_name), 'a+') as outfile:
+                            outfile.write('\n')
+                            json.dump(event,outfile)
+                            self.successfully_sent_events_number += 1
+                except Exception as err:
+                    helper.log_error('Error while writing data to buffer in bin. {}'.format(err))
+                
+            else :
+                #non-durable ingestion mode where the incoming events are directly ingested to ADX
+                retry_count = 0
+                while retry_count < self.retry_times:
+                    try:
+                        event_str = json.dumps(event)
+                        json_bytes = event_str.encode('utf-8')
+                        stream = io.BytesIO(json_bytes)
+                        stream_descriptor = StreamDescriptor(stream)
+                        self.client.ingest_from_stream(stream_descriptor, ingestion_properties=self.ingestion_props)
+                        self.successfully_sent_events_number += 1
+                        break # Event posted to ADX, break out of the retry loop
+                    except KustoServiceError as ex:
```

Comment:
Permanent vs Transient error checks like we do in other apps, should we do it

Reply:
done, added in the next commit